File: src/core/Cryptos/CryptoTransaction.cs
using System;
using core.Shared;

namespace core.Cryptos
{
    public class CryptoTransaction
    {
        public Guid AggregateId;
        public Guid TransactionId;
        public string Token;
        public string Description;
        public decimal Price;
        public decimal Debit;
        public decimal Credit;
        public DateTimeOffset When;

        public CryptoTransaction(Guid aggregateId, Guid transactionId, string token, string description, decimal price, decimal debit, decimal credit, DateTimeOffset when)
        {
            AggregateId = aggregateId;
            TransactionId = transactionId;
            Token = token;
            Description = description;
            Price = price;
            Debit = debit;
            Credit = credit;
            When = when;
        }

        internal static CryptoTransaction DebitTx(Guid aggregateId, Guid transactionId, string token, string description, decimal price, decimal dollarAmount, DateTimeOffset when)
        {
            return new CryptoTransaction(
                aggregateId,
                transactionId,
                token,
                description,
                price,
                debit: dollarAmount,
                credit: 0,
                when
            );
        }

        internal static CryptoTransaction CreditTx(Guid aggregateId, Guid transactionId, string token, string description, decimal price, decimal dollarAmount, DateTimeOffset when)
        {
            return new CryptoTransaction(
                aggregateId,
                transactionId,
                token,
                description,
                price: price,
                debit: 0,
                credit: dollarAmount,
                when
            );
        }

        public Transaction ToSharedTransaction() => Credit switch {
            > 0 => Transaction.NonPLTx(AggregateId, TransactionId, new Ticker(Token), Description, Price, Credit, When, isOption: false),
            _ => Transaction.NonPLTx(AggregateId, TransactionId, new Ticker(Token), Description, Price, Debit, When, isOption: false)
        };
    }
}

File: src/core2.fs/Cryptos/CryptoTransaction.fs
namespace core.Cryptos

open System
open core.Shared

[<AllowNullLiteral>]
type CryptoTransaction(aggregateId: Guid, transactionId: Guid, token: string, description: string, price: decimal, debit: decimal, credit: decimal, ``when``: DateTimeOffset) =
    member val AggregateId = aggregateId with get
    member val TransactionId = transactionId with get
    member val Token = token with get
    member val Description = description with get
    member val Price = price with get
    member val Debit = debit with get
    member val Credit = credit with get
    member val When = ``when`` with get

    static member internal DebitTx(aggregateId: Guid, transactionId: Guid, token: string, description: string, price: decimal, dollarAmount: decimal, ``when``: DateTimeOffset) =
        CryptoTransaction(
            aggregateId,
            transactionId,
            token,
            description,
            price,
            dollarAmount,
            0m,
            ``when``
        )

    static member internal CreditTx(aggregateId: Guid, transactionId: Guid, token: string, description: string, price: decimal, dollarAmount: decimal, ``when``: DateTimeOffset) =
        CryptoTransaction(
            aggregateId,
            transactionId,
            token,
            description,
            price,
            0m,
            dollarAmount,
            ``when``
        )

    member this.ToSharedTransaction() =
        match this.Credit with
        | c when c > 0m -> Transaction.NonPLTx(this.AggregateId, this.TransactionId, Ticker(this.Token), this.Description, this.Price, this.Credit, this.When, isOption = false)
        | _ -> Transaction.NonPLTx(this.AggregateId, this.TransactionId, Ticker(this.Token), this.Description, this.Price, this.Debit, this.When, isOption = false)
